### ISSUE TITLE
Add support for React 18

### ISSUE THREAD

**Issue by GrossDesignCo** [storybookjs/storybook#17831]
**Is your feature request related to a problem? Please describe**

Only so far as upgrading to React 18 is concerned :)

Specifically, when I try to update my `devDependencies`, `@storybook/addon-controls` throws an error, because it's `peerDependencies` is set to only React 16 and 17.

Doublechecked in the package.json
https://github.com/storybookjs/storybook/blob/next/addons/controls/package.json#L61-L64

There's probably already some awareness of this and/or a fix in flight, but just in case I thought I'd mention it.

**Describe the solution you'd like**

Ideally I think it's as simple as updating the `peerDependencies` throughout Storybook's packages, but I'm sure there's some additional work in the ReactDOM.render > ReactDOM.createRoot transition that would affect Storybook's rendering cycle.

**Describe alternatives you've considered**

None as yet

**Are you able to assist to bring the feature to reality?**

I'm happy to make a PR to update the peerDeps, if it was deeper than that I'd be lost :)

**Comment by souppower** [comment]
@GrossDesignCo 

This pr might be related.
https://github.com/storybookjs/storybook/pull/17215

It doesn't look like it's as simple as bumping the peerDependencies. 🤔

**Comment by bcheidemann** [comment]
It would be great to get some indication from the storybook team about when storybook will support React 1. Many projects will be unable to upgrade until storybook has support.

**Comment by cameron-martin** [comment]
It should be as simple as bumping the peerDependencies if you don't want storybook to use the new `createRoot` APIs. The old `ReactDOM.render` API should behave exactly the same AFAIK.

**Comment by shilman** [comment]
I'm hoping to get https://github.com/storybookjs/storybook/pull/17215 merged and released early next week. If that doesn't happen for some reason, I'll look into a simpler PR to start to bump the peer dependencies.

**Comment by shilman** [comment]
Olé!! I just released https://github.com/storybookjs/storybook/releases/tag/v6.5.0-alpha.54 containing PR #17853 that references this issue. Upgrade today to the `@next` NPM tag to try it out!

```
npx sb upgrade --prerelease
```

Closing this issue. Please re-open if you think there's still more to do.

**Comment by jcerri** [comment]
@shilman I couldn't verify if this fixes the issue I still got but it may be needed to do the same for react-test-renderer dep.

**Comment by shilman** [comment]
Ermahgerd!! I just released https://github.com/storybookjs/storybook/releases/tag/v6.5.0-alpha.58 containing PR #17215 that references this issue. Upgrade today to the `@next` NPM tag to try it out!

```
npx sb upgrade --prerelease
```

Closing this issue. Please re-open if you think there's still more to do.

**Comment by brendonco** [comment]
Getting these warnings after upgrade @shilman 

```
WARN Found 7 outdated packages (relative to '@storybook/addon-actions@6.5.0-alpha.58')
WARN Please make sure your packages are updated to ensure a consistent experience.
WARN - @storybook/node-logger@6.4.20
WARN - @storybook/addon-postcss@2.0.0
WARN - @storybook/react-docgen-typescript-plugin@1.0.2-canary.6.9d540b91e815f8fc2f8829189deb00553559ff63.0
WARN - @storybook/babel-plugin-require-context-hook@1.0.1
WARN - @storybook/mdx1-csf@0.0.1-canary.1.867dcd5.0
WARN - @storybook/mdx1-csf@0.0.1-canary.1.867dcd5.0
WARN - @storybook/mdx2-csf@0.0.1-canary.1.838a6ca.0
```

**Comment by shilman** [comment]
Thanks @brendonco -- safe to ignore and I'll fix in the next release

**Comment by JohnCido** [comment]
I'm getting this after upgrading to 6.5.0-alpha.59:
```console
The pseudo class ":first-child" is potentially unsafe when doing server-side rendering. Try changing it to ":first-of-type".
```
Logs attached:
[localhost-1649323648198.log](https://github.com/storybookjs/storybook/files/8441571/localhost-1649323648198.log)
[localhost-1649323685702.log](https://github.com/storybookjs/storybook/files/8441572/localhost-1649323685702.log)

**Comment by shilman** [comment]
@valentinpalkovic any ideas? ☝️

**Comment by valentinpalkovic** [comment]
@JohnCido Which version have you used before?

@shilman I assume this warning is thrown by emotion und should be unrelated of the react 18 support. I rather think, that in the meantime emotion was potentially updated, too.

Related issues:
- https://github.com/emotion-js/emotion/issues/1105

Worth to mention: The message gets already ignored in jest tests https://github.com/storybookjs/storybook/blob/3f3e65001ff6dadebf4e753640a7d94c40b3285e/scripts/jest.init.js#L36 for quite some time.

**Comment by JohnCido** [comment]
> @JohnCido Which version have you used before? And can you tell me, which addons are in use?
> 
> @shilman I assume this warning is thrown by emotion und should be unrelated of the react 18 support. I rather think, that in the meantime emotion was potentially updated, too.
> 
> Related issues:
> 
> * [Cannot turn off "potentially unsafe when doing server-side rendering" noise emotion-js/emotion#1105](https://github.com/emotion-js/emotion/issues/1105)
> 
> Worth to mention: The message gets already ignored in jest tests
> 
> https://github.com/storybookjs/storybook/blob/3f3e65001ff6dadebf4e753640a7d94c40b3285e/scripts/jest.init.js#L36
> 
> for quite some time.
> We could the required `ignoreSSrFlag` to the remaining usages of `first-child` to avoid the warning. Like it is done here:
> 
> https://github.com/storybookjs/storybook/blob/783e5aaf26dc4c5312133dffca29caa543abad7c/lib/components/src/blocks/ArgsTable/ArgsTable.tsx#L114
> 
> Indeed, an adjustment would only be necessary in `lib/components/src/blocks/Source.tsx:L55`.
> 
> I will open a PR

@valentinpalkovic 
I was using alpha 57 before this upgrade. And these are the `@storybook/` packages I've been using:
```JSON
"@storybook/addon-actions": "^6.5.0-alpha.59",
"@storybook/addon-essentials": "^6.5.0-alpha.59",
"@storybook/addon-interactions": "^6.5.0-alpha.59",
"@storybook/addon-links": "^6.5.0-alpha.59",
"@storybook/api": "^6.5.0-alpha.59",
"@storybook/builder-webpack5": "^6.5.0-alpha.59",
"@storybook/manager-webpack5": "^6.5.0-alpha.59",
"@storybook/react": "^6.5.0-alpha.59",
```

**Comment by Hypnosphi** [comment]
There's still a "Warning: ReactDOM.render is no longer supported in React 18" in `6.5.0-alpha.59` because manager still uses ReactDOM.render: https://github.com/storybookjs/storybook/blob/next/lib/ui/src/index.tsx#L91

**Comment by valentinpalkovic** [comment]
@Hypnosphi Thank you for the hint. This issue is already known (discussed with @shilman) and we decided to take care of the new Root API for the Storybook UI in an upcoming iteration.

**Comment by shilman** [comment]
Boo-yah!! I just released https://github.com/storybookjs/storybook/releases/tag/v6.5.0-alpha.60 containing PR #17911 that references this issue. Upgrade today to the `@next` NPM tag to try it out!

```
npx sb upgrade --prerelease
```

**Comment by bonesoul** [comment]
also looking for this.

**Comment by valentinpalkovic** [comment]
@alexNerazim, @bonesoul Can you tell me which versions of the storybook dependencies are installed?

**Comment by alexNerazim** [comment]
@valentinpalkovic I have this versions:

   "@storybook/addon-actions": "^6.4.22",
    "@storybook/addon-essentials": "^6.4.22",
    "@storybook/addon-interactions": "^6.4.22",
    "@storybook/addon-links": "^6.4.22",
    "@storybook/builder-vite": "^0.1.28",
    "@storybook/testing-library": "^0.0.9",
    "eslint-plugin-storybook": "^0.5.8",

**Comment by valentinpalkovic** [comment]
As described, you have to use the latest preview version of Storybook. Otherwise, React 18 will not work together with Storybook. The latest prerelease alpha version is `v6.5.0-alpha.64`

**Comment by valentinpalkovic** [comment]
@MarkiyanPyts

The „Warning: ReactDOM.render is no longer supported in React 18“ warning in the console is expected, because the storybook builder still uses the old React Root API. 

If you still have some issues, please provide a reproducible example. https://storybook.js.org/docs/react/contribute/how-to-reproduce

**Comment by alexNerazim** [comment]
I did the update to the latest version and still having issues.
Storybook stay loading all the time.
This are the issues that appear in the console:

<img width="950" alt="image" src="https://user-images.githubusercontent.com/99650672/164220174-f08602d6-e169-46c6-ad8f-60957f65d548.png">

**Comment by valentinpalkovic** [comment]
Try to delete the yarn.lock/package-lock file and delete the node_modules and trigger an installation one more time. Maybe some build assets are cached. If this doesn‘t help, please create a reproduction example like explained above. Then I can try to help you.

**Comment by FuriouZz** [comment]
@valentinpalkovic In my case I am a fresh start

```sh
npx create-react-app . --template typescript
npx sb init --type react --use-npm
```

**Comment by Hypnosphi** [comment]
Also, storyshots still depend on react-test-renderer 16 || 17

**Comment by shilman** [comment]
@Hypnosphi https://github.com/storybookjs/storybook/pull/17865

**Comment by ctsstc** [comment]
I tried: `v6.5.0-beta.1` and `v6.5.0-alpha.64` with no luck.

**Comment by sjoqvist** [comment]
@ctsstc This looks more like a Node 18 than a React 18 problem. Have you tried to downgrade Node?

I'm using Node 16.14.2 and I get exactly the same error as you when upgrading to Node 18.0.0 (without upgrading from Storybook 6.4.21). Upgrading to Storybook 6.5 beta 1 without upgrading Node still gives me an error (which I'll see if I can understand better) but it occurs much later in the build process and seems to be unrelated.

**Comment by shilman** [comment]
@ctsstc @sjoqvist mind opening a new issue for this one? would be great to squeeze node18 support into 6.5

**Comment by sjoqvist** [comment]
@shilman Looking at the existing issues, I believe that this is already covered by #18019 (although the workaround mentioned there doesn't work for me).

**Comment by shilman** [comment]
Good call @sjoqvist 🙏 hopefully they get it fixed soon--seems like it's affecting a lot of people

**Comment by pixelass** [comment]
There's an issue with MDX:

Most likely related to `@mdx-js/react`

The list of errors is long. Mostly related to the following.
```
[…]
 Field 'browser' doesn't contain a valid alias configuration
              /Users/*********/node_modules/@mdx-js/react.ts doesn't exist
[…]
```

storybook uses: "@mdx-js/react": "^1.6.22"
latest is: "@mdx-js/react": "2.1.1"

After installing the latest version of `@mdx-js/react` new errors pop up since it has breaking changes.

@shilman Sorry but I won't be able to create a reproduction for this issue. I understand how helpful that would be but my time is very limited ATM.

Potential workarounds: 
* Use React 17 (💩)

**Comment by shilman** [comment]
@pixelass  have you tried opting MDXv2 in 6.5?

https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#opt-in-mdx2-support

**Comment by BiancaArtola** [comment]
After migrating to 6.5.4 I am still having the same issue 😢

**Comment by TomasKomprs** [comment]
> ```shell
> grep version node_modules/ip/package.json
> ```

same

**Comment by shilman** [comment]
@aviyi @TodBob can you please elaborate?

**Comment by TodBob** [comment]
> @aviyi @TodBob can you please elaborate?

Still getting Warning in the console: 
" Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot"

Also new error, but I don't know yet if it is connected with the storybook:
"The pseudo class ":first-child" is potentially unsafe when doing server-side rendering. Try changing it to ":first-of-type"."

I have:
react v18.1.0
storybook v6.5.6

**Comment by valentinpalkovic** [comment]
" Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot" -> if you don't see this error twice, than everything is fine. Storybooks Manager isn't rewritten so far and will use the legacy root API to mount itself to the DOM. You can safely ignore this message.

**Comment by TodBob** [comment]
> `Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot` -> if you don't see this error twice, then everything is fine. Storybooks Manager isn't rewritten so far and will use the legacy root API to mount itself to the DOM. You can safely ignore this message.

I see it only once. So it's OK, thanks.

**Comment by sjoqvist** [comment]
> ```shell
> grep version node_modules/ip/package.json
> ```
> 
> same

That wasn't supposed to fix anything. I wanted to know which version of `node-ip` you have installed. That command should return the version, so you'll need to paste the output.

**Comment by sarujo** [comment]
> > @aviyi @TodBob can you please elaborate?
> 
> Still getting Warning in the console: " Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: [reactjs.org/link/switch-to-createroot](https://reactjs.org/link/switch-to-createroot)"
> 
> Also new error, but I don't know yet if it is connected with the storybook: "The pseudo class ":first-child" is potentially unsafe when doing server-side rendering. Try changing it to ":first-of-type"."
> 
> I have: react v18.1.0 storybook v6.5.6

Same here

**Comment by jonesmac** [comment]
> > > @aviyi @TodBob can you please elaborate?
> > 
> > 
> > Still getting Warning in the console: " Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: [reactjs.org/link/switch-to-createroot](https://reactjs.org/link/switch-to-createroot)"
> > Also new error, but I don't know yet if it is connected with the storybook: "The pseudo class ":first-child" is potentially unsafe when doing server-side rendering. Try changing it to ":first-of-type"."
> > I have: react v18.1.0 storybook v6.5.6
> 
> Same here

See https://github.com/storybookjs/storybook/issues/17831#issuecomment-1143670115

You should only be concerned if you see it twice.  I think this is a little odd to claim React 18 support but then expect people to know its working if a console error only fires twice.

**Comment by juandgarin** [comment]
Hello any update?

**Comment by sjoqvist** [comment]
> Hello any update?

Can you confirm that you don't have `node-ip` below 1.1.6 installed?

**Comment by swapnonil-b** [comment]
When I was at 6.5.0~beta my docgen was working, but after I upgraded to stable versions (6.5.5+) it does not automatically generate argtypes.

**Comment by agrauch** [comment]
@pixelass another work around is to use `npm install --legacy-peer-deps`. Not recommended but...

**Comment by khats** [comment]
@shilman should manager also be rendered in react18 mode or no?
as i see there is still
```
ReactDOM.render
```
https://github.com/storybookjs/storybook/blob/v6.5.9/lib/core-client/src/manager/index.ts#L10

https://github.com/storybookjs/storybook/blob/v6.5.9/lib/ui/src/index.tsx#L96

**Comment by valentinpalkovic** [comment]
@khats https://github.com/storybookjs/storybook/issues/17831#issuecomment-1143670115

**Comment by saitonakamura** [comment]
@valentinpalkovic do I understand correctly that warning from [renderDocs](https://github.com/storybookjs/storybook/blob/b4bf8262bd1d9be367bf240ba407cac4b19c9ea2/lib/preview-web/src/renderDocs.tsx#L44) is also ok for now? Since main part is [here](https://github.com/storybookjs/storybook/blob/b4bf8262bd1d9be367bf240ba407cac4b19c9ea2/app/react/src/client/preview/render.tsx#L39-L46) ?

**Comment by valentinpalkovic** [comment]
Yes, should be fine as well for now. :)

**Comment by atomless** [comment]
Storybook was running fine with React 17. It stopped working after the React 18 upgrade with the warning described in this issues above. After running the proposed fix 
```
npx sb upgrade --prerelease
```
I now see the error: `start-storybook: command not found`

It's like the upgrade nuked storybook completely

**Comment by saitonakamura** [comment]
@atomless you no longer need upgrade to prerelease, it's in the vanilla release right now, that fix is outdated

**Comment by atomless** [comment]
thank you @saitonakamura - sadly, I started on latest (6.5.9) and it is on that version that storybook starts ok but none of the stories render, plus in dev tools I see the warning `ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17.`

**Comment by saitonakamura** [comment]
@atomless some of this may be fine, you may want to check out last couple of messages in this thread

**Comment by atomless** [comment]
@saitonakamura yeah looks like I'm seeing the same issue as @Kestutis

**Comment by saitonakamura** [comment]
@atomless than this is fine, look over there https://github.com/storybookjs/storybook/issues/17831#issuecomment-1174736520

**Comment by atomless** [comment]
@saitonakamura thanks again but it appears my issue is along a different path of execution through the core of 6.5.9
this is the console log stack:
![Storybook-render3](https://user-images.githubusercontent.com/28726/181211165-a45d142f-1131-4ce0-98a4-9bd66d412b26.png)
and here's the point it calls the method that throws:
![Storybook-render2](https://user-images.githubusercontent.com/28726/181211341-8db7a0fb-3ee5-4e21-b23c-5aaf78fd5a76.png)
and the method itself:
![Storybook-render](https://user-images.githubusercontent.com/28726/181211515-8098d9c0-d5e7-4fa7-a96e-afc819cb043e.png)

**Comment by khats** [comment]
@atomless  manager panel is still renders with old method as @valentinpalkovic answered here: https://github.com/storybookjs/storybook/issues/17831#issuecomment-1143670115
do you have any others errors in console? can you provide part of your `main.js` file for storybook setup?

**Comment by atomless** [comment]
Aha! Thanks @khats. here you go : 
```
module.exports = {  
  "core": {
    "builder": "webpack5",
  },
  "stories": [
    "../storybook/stories/*.stories.mdx",
    "../storybook/stories/*.stories.@(js|jsx|ts|tsx)"
  ],
  "addons": [
    "storybook-preset-craco",
    "storybook/preset-create-react-app",
    "@storybook/addon-essentials",
    "@storybook/addon-links"
  ]
}```

**Comment by khats** [comment]
@atomless i am sure(maybe some guru will correct) me, but in my config `core` defined like this:
```
module.exports = {  
  "core": {
    "builder": {
         name: "webpack5",
    }
    ....
  },
  ```
and what do you have in terminal? 
my example:
```
info @storybook/react v6.5.9
info 
info => Loading presets
info => Loading custom manager config
info => Using PostCSS preset with postcss@8.4.5
info => Using React fast refresh
info => Using default Webpack5 setup

```
<img width="738" alt="image" src="https://user-images.githubusercontent.com/971413/181223979-634a1cb2-c623-4f9c-b0e7-3508e36f254d.png">

**Comment by atomless** [comment]
@khats - interesting - I just noticed an error : 
```info @storybook/react v6.5.9ing npm:load Completed in 37ms
info 
info => Loading presets: timing npm:load Completed in 37ms
WARN   Failed to load preset: {"type":"presets"} on level 1
```
and I see no request for `stories.json` which possibly explains why I just see a spinner and an otherwise blank storybook ui/sidepanel

**Comment by khats** [comment]
@atomless i have found more or less similar issue, maybe it will help to you
https://github.com/storybookjs/storybook/issues/7055

**Comment by atomless** [comment]
@khats - thanks but following the solution there just caused way more errors in the terminal than before 🤔

**Comment by atomless** [comment]
@valentinpalkovic know of a work around for manager panel continuing to use the old react render method as it's the only error I see now when running Storybook and all I see is the empty ui with a spinner, no stories ever render??

**Comment by valentinpalkovic** [comment]
@atomless Please open a new issue and post a reproduction example. You can mention me, and I will try to help you.

**Comment by GuillaumeCisco** [comment]
> `Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot` -> if you don't see this error twice, then everything is fine. Storybooks Manager isn't rewritten so far and will use the legacy root API to mount itself to the DOM. You can safely ignore this message.

Do we know when it will be finally fixed?
Do we have an ETA?

**Comment by alberto-f** [comment]
> `Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot` -> if you don't see this error twice, then everything is fine. Storybooks Manager isn't rewritten so far and will use the legacy root API to mount itself to the DOM. You can safely ignore this message.

link: https://github.com/storybookjs/storybook/issues/17831#issuecomment-1143670115

Would it be possible to get some more information about the sentence `if you don't see this error twice, then everything is fine` ? If I see the error once, Are my components using react v18 or react v17 ?

Is there any ticket to get information about when storybook will fully support react v18  as this one has been closed ?

This thread provides hacks to move forward and use react 18 but it is soooo lengthy that it is difficult to follow.

If storybook does not support yet react v18, guys, people, IT IS ALL GOOD. You guys are doing an amazing job and the fact that many people are asking in this thread is because we love storybook and want to keep it in our projects. But a more official answer would help so that everyone does not need to read this thread and still be unsure if storybook does support it without warnings or hacks such as using overrides in package.json or by using --legacy-peer-deps (based on this one https://stackoverflow.com/a/66620869/1578841 it does not seem to me really recommended to follow as it mutes valid checks and warnings)

**Comment by valentinpalkovic** [comment]
@alberto-f Thank you for your concerns. 

I try to do my best to summarize everything important to the React 18 support.

1. Storybook will render your components with the [new root API](https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-client-rendering-apis) as soon as React 18 is installed. You can [opt-out](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#react18-new-root-api) from this if you want. Then the legacy root API is used to mount your components
2. Storybook itself, the Storybook manager, is written in React, too. And it relies on the React version, which is installed in your project. It doesn't provide its very own React version. Therefore, if React 18 is used in your project, the Storybook Manager uses it as well, but will still use the legacy root API to mount itself to the DOM. This mounting has nothing to do with how your components are mounted to the DOM.

**Next steps**

As discussed with the maintainers, we didn't want to use the new Root API to mount Storybook to the DOM because this would require profound testing of Storybook and all of its plugins. I am sure this will be done in one of the next iterations. Maybe @shilman could give some feedback in this regard.

**Questions**

> If I see the error once, Are my components using react v18 or react v17 ?

Which React version your components use relies on your installed React version. If you are installing React in version 18, then version 18 is used. The difference is only whether the concurrent features are activated. And this depends of whether the components are mounted with the new Root API or the legacy one. As already mentioned, if React 18 is in use and you haven't opt-out from the new Root API, the Root API is used, and all concurrent features are "activated".

> Is there any ticket to get information about when Storybook will fully support react v18 as this one has been closed?

Storybook itself supports the mounting of your components in React 18 with the new Root API. If still, some errors occur, please open a new ticket with a reproducible example. Often, the issue was related to an existing yarn.lock file, which didn't allow to resolve dependencies correctly. In some other cases, some transitive dependencies were not ready for React 18. Ideally, the support should be fine (except for the console error message from React, which tells you, that the Storybook Manager still uses the old API)

> This thread provides hacks to move forward and use react 18, but it is soooo lengthy that it is difficult to follow.

Hacks shouldn't be needed at all. Please create a new ticket if you still have any issues (except for the console error from React). I am sure it will be taken care of as soon as possible.

**Comment by alberto-f** [comment]
@valentinpalkovic Thank you so much for taking the time to clarify all the questions.

It is really clear the status of storybook supporting React v18 and how to proceed if something is not working as expected.

**Comment by stephyswe** [comment]
> @alberto-f Thank you for your concerns.
> 
> I try to do my best to summarize everything important to the React 18 support.
> 
> 1. Storybook will render your components with the [new root API](https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-client-rendering-apis) as soon as React 18 is installed. You can [opt-out](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#react18-new-root-api) from this if you want. Then the legacy root API is used to mount your components
> 2. Storybook itself, the Storybook manager, is written in React, too. And it relies on the React version, which is installed in your project. It doesn't provide its very own React version. Therefore, if React 18 is used in your project, the Storybook Manager uses it as well, but will still use the legacy root API to mount itself to the DOM. This mounting has nothing to do with how your components are mounted to the DOM.
> 
> **Next steps**
> 
> As discussed with the maintainers, we didn't want to use the new Root API to mount Storybook to the DOM because this would require profound testing of Storybook and all of its plugins. I am sure this will be done in one of the next iterations. Maybe @shilman could give some feedback in this regard.
> 
> **Questions**
> 
> > If I see the error once, Are my components using react v18 or react v17 ?
> 
> Which React version your components use relies on your installed React version. If you are installing React in version 18, then version 18 is used. The difference is only whether the concurrent features are activated. And this depends of whether the components are mounted with the new Root API or the legacy one. As already mentioned, if React 18 is in use and you haven't opt-out from the new Root API, the Root API is used, and all concurrent features are "activated".
> 
> > Is there any ticket to get information about when Storybook will fully support react v18 as this one has been closed?
> 
> Storybook supports mounting your components in React 18 with the new Root API. If still, some errors occur, please open a new ticket with a reproducible example. Often, the issue was related to an existing yarn.lock file, which didn't allow to resolve dependencies correctly. In some other cases, some transitive dependencies were not ready for React 18. Ideally, the support should be fine (except for the console error message from React, which tells you, that the Storybook Manager still uses the old API)
> 
> > This thread provides hacks to move forward and use react 18, but it is soooo lengthy that it is difficult to follow.
> 
> Hacks shouldn't be needed at all. I know that npm users might have the `legacy-peer-deps` issue because some transitive dependencies haven't defined React 18 as a valid React version. That's indeed awkward. Please create a new ticket if you still have any issues (except for the console error from React). I am sure it will be taken care of as soon as possible.

Take your time. Only thing that is the issue is this little warning message. When we find a solid way to get rid of that message (aka upgrade something). I think its better to wait until that fix arrives. :)

**Comment by NathanVss** [comment]
> @alberto-f Thank you for your concerns.
> 
> I try to do my best to summarize everything important to the React 18 support.
> 
> 1. Storybook will render your components with the [new root API](https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-client-rendering-apis) as soon as React 18 is installed. You can [opt-out](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#react18-new-root-api) from this if you want. Then the legacy root API is used to mount your components
> 2. Storybook itself, the Storybook manager, is written in React, too. And it relies on the React version, which is installed in your project. It doesn't provide its very own React version. Therefore, if React 18 is used in your project, the Storybook Manager uses it as well, but will still use the legacy root API to mount itself to the DOM. This mounting has nothing to do with how your components are mounted to the DOM.
> 
> **Next steps**
> 
> As discussed with the maintainers, we didn't want to use the new Root API to mount Storybook to the DOM because this would require profound testing of Storybook and all of its plugins. I am sure this will be done in one of the next iterations. Maybe @shilman could give some feedback in this regard.
> 
> **Questions**
> 
> > If I see the error once, Are my components using react v18 or react v17 ?
> 
> Which React version your components use relies on your installed React version. If you are installing React in version 18, then version 18 is used. The difference is only whether the concurrent features are activated. And this depends of whether the components are mounted with the new Root API or the legacy one. As already mentioned, if React 18 is in use and you haven't opt-out from the new Root API, the Root API is used, and all concurrent features are "activated".
> 
> > Is there any ticket to get information about when Storybook will fully support react v18 as this one has been closed?
> 
> Storybook supports mounting your components in React 18 with the new Root API. If still, some errors occur, please open a new ticket with a reproducible example. Often, the issue was related to an existing yarn.lock file, which didn't allow to resolve dependencies correctly. In some other cases, some transitive dependencies were not ready for React 18. Ideally, the support should be fine (except for the console error message from React, which tells you, that the Storybook Manager still uses the old API)
> 
> > This thread provides hacks to move forward and use react 18, but it is soooo lengthy that it is difficult to follow.
> 
> Hacks shouldn't be needed at all. I know that npm users might have the `legacy-peer-deps` issue because some transitive dependencies haven't defined React 18 as a valid React version. That's indeed awkward. Please create a new ticket if you still have any issues (except for the console error from React). I am sure it will be taken care of as soon as possible.

Thank you for this precision, it prevented us to waste time in trying to find a solution to this error message. The main point to keep in mind for us is that **Storybook renders of components with React 18** ( if installed ), so when we interact with them we are "really" testing them with React 18 and its concurrent features.

Thank you for your amazing work :)

**Comment by vislogurov** [comment]
@atomless Yo! DId you solve your problem with "ui with a spinner, no stories ever render"? I have the same problem too(

**Comment by edgar0011** [comment]
This worked for me:
`import React from 'react'

window.React = React`
